ISO8601::parseCalendarTime must throw in cases of time/date ambiguity

rkirsling opened the pull request:
#### 12985f546f1e089188516f52cea8d71c0605d446
<pre>
ISO8601::parseCalendarTime must throw in cases of time/date ambiguity
<a href="https://bugs.webkit.org/show_bug.cgi?id=240498">https://bugs.webkit.org/show_bug.cgi?id=240498</a>

Reviewed by Yusuke Suzuki.

The CalendarTime production (<a href="https://tc39.es/proposal-temporal/#prod-CalendarTime)">https://tc39.es/proposal-temporal/#prod-CalendarTime)</a> goes to a lot of trouble
to ensure that ambiguous parses are not allowed; this patch ensures that our implementation does so too.

In short:
- CalendarTime is basically TimeDesignator[opt] TimeSpec TimeZone[opt] Calendar[opt].
- When we don&apos;t have a TimeDesignator or a Calendar, we&apos;re in danger of ambiguity:
  YYYY-MM, YYYYMM, MM-DD, MMDD look the same as HHMM-UU, HHMMSS, HH-UU, HHMM (where UU is a UTC offset).
- The solution is to attempt to reparse the string as one of these date formats and throw if we succeed.

* stress/temporal-plaintime.js: Adjust expectations.
* test262/expectations.yaml: Mark 10 test cases as passing.

* runtime/ISO8601.cpp:
(JSC::ISO8601::isAmbiguousCalendarTime):
(JSC::ISO8601::parseCalendarTime):
(JSC::ISO8601::daysInMonth):
* runtime/ISO8601.h:

Canonical link: <a href="https://commits.webkit.org/250634@main">https://commits.webkit.org/250634@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@294306">https://svn.webkit.org/repository/webkit/trunk@294306</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1154,9 +1185,15 @@ static constexpr uint8_t daysInMonths[2][12] = {
+}
+
+uint8_t daysInMonth(uint8_t month)
+{
```
Let's define
```
constexpr unsigned isLeapYear = 1;
return daysInMonths[isLeapYear][month - 1];
```
to make this code easier to read.

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1039,14 +1014,70 @@ std::optional<std::tuple<PlainTime, std::optional<TimeZoneRecord>>> parseTime(St
 
+static bool isAmbiguousCalendarTime(StringView string)
+{
+    // String is known to be valid as `TimeSpec TimeZone[opt]`, so DateExtendedYear and TwoDashes are not possible.
```
Let's use readCharactersForParsing in this function to make it more efficient and aligned to the rest of ISO8601.

```
return readCharactersForParsing(string, [](auto buffer) -> bool {
    unsigned length = buffer.lengthRemaining();
    ASSERT(length);
    if (buffer[0] == 'T' || buffer[0] == 't')
        return false;
    ASSERT(isASCIIDigit(buffer[0]) && isASCIIDigit(buffer[1]));
    ...
});
```

And we can replace `string` with `buffer`.

Constellation reviewed (changes requested):
Looks great! Some nits.

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1039,14 +1014,70 @@ std::optional<std::tuple<PlainTime, std::optional<TimeZoneRecord>>> parseTime(St
+
+    // Without an explicit T prefix or a calendar, we need to verify that the parse isn't ambiguous with DateSpecYearMonth or DateSpecMonthDay.
+    if (tuple && !std::get<2>(tuple.value()) && string[0] != 'T' && string[0] != 't') {
+        if (isAmbiguousCalendarTime(string))
```
We can integrate `'t'` and `'T'` check into `isAmbiguousCalendarTime` too.

```
if (tuple && !std::get<2>(tuple.value())) {
    if (isAmbiguousCalendarTime(string))
        return std::nullopt;
}
```

Constellation reviewed (approved):
r=me

zdobersek committed:
[GTK][WPE] Current-context enforcement in ANGLE is expensive
https://bugs.webkit.org/show_bug.cgi?id=240392

Reviewed by Adrian Perez de Castro.

Source/WebCore:

ANGLE's current-context operations are relatively heavy and introduce
detectable overhead when executed for each of many WebGL operation. More
specifically, loads of the current-thread data from thread-local
storage are done under both EGL_GetCurrentContext() and
EGL_MakeCurrent() through the general TLS access model, relying on the
__tls_get_addr function.

Instead, a static thread-local pointer variable is introduced, with the
initial-exec method, containing the address of the current
GraphicsContextGLANGLE object for a given thread. Instead of dipping
into ANGLE to retrieve the current context, the GraphicsContextGLANGLE
object's address is compared to the value of the thread-local variable
and an early return is done if they match.

This model is applied through the TLS_MODEL_INITIAL_EXEC macro that's
defined appropriately if the compiler is detected to support the
required tls_model attribute. This should be the case when targetting
platforms that use ELF, i.e. Unix-like systems but primarily Linux.

The initial-exec TLS model relies on early understanding from the
dynamic linker about what TLS data is required, which enables more
efficient allocation of and access into the required underlying memory.

Caching the current-context data into a static thread-specific variable
with the initial-exec TLS model is the same approach that's taken in the
Mesa library. More on this and other TLS models can be gathered from the
'ELF Handling For Thread-Local Storage' paper by Ulrich Drepper.

* platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLANGLE::makeContextCurrent):

Source/WTF:

* wtf/Compiler.h: Define the TLS_MODEL_INITIAL_EXEC macro as the
tls_model attribute with the initial-exec value, to be used on variables
whose TLS model should be adjusted.

Canonical link: https://commits.webkit.org/250632@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@294290 268f45cc-cd09-0410-ab3c-d52691b4dbfc

anttijk committed:
Resolve ::first-letter eagerly
https://bugs.webkit.org/show_bug.cgi?id=239844

Reviewed by Alan Bujtas.

Source/WebCore:

::first-letter pseudo-element should be resolved eagerly during style resolution, similar to ::first-line.
This will clean up the architecture. It also makes ::first-letter animatable.

* rendering/RenderTextFragment.cpp:
(WebCore::RenderTextFragment::styleDidChange): Deleted.
* rendering/RenderTextFragment.h:
* rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::removeCachedPseudoStyle): Deleted.

We always create a new style. There is no need to removed pseudo element styles.

* rendering/style/RenderStyle.h:
* rendering/updating/RenderTreeBuilderFirstLetter.cpp:
(WebCore::styleForFirstLetter):

Pull first-letter style from the parent style instead of resolving it in the middle of render tree building.

(WebCore::RenderTreeBuilder::FirstLetter::updateAfterDescendants):
(WebCore::RenderTreeBuilder::FirstLetter::updateStyle):

Delete the renderer if we no longer have a style.

(WebCore::RenderTreeBuilder::FirstLetter::createRenderers):
* rendering/updating/RenderTreeBuilderFirstLetter.h:
* style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):

Resove first-letter.

(WebCore::Style::TreeResolver::resolvePseudoElement):
(WebCore::Style::TreeResolver::resolveInheritedPseudoElement):

Factor into a function.

(WebCore::Style::isChildInBlockFormattingContext):
(WebCore::Style::TreeResolver::resolveAncestorFirstLinePseudoElement):
(WebCore::Style::TreeResolver::resolveAncestorFirstLetterPseudoElement):

Resolving first-letter is similar to first-line except it is always a leaf and so won't be inherited from.

(WebCore::Style::TreeResolver::makeResolutionContextForPseudoElement):
(WebCore::Style::TreeResolver::makeResolutionContextForInheritedFirstLine):
* style/StyleTreeResolver.h:

LayoutTests:

* fast/css-generated-content/first-letter-in-nested-before-expected.html: Added.
* fast/css-generated-content/first-letter-in-nested-before-expected.png: Removed.
* fast/css-generated-content/first-letter-in-nested-before-expected.txt: Removed.
* fast/css-generated-content/first-letter-in-nested-before-table-expected.html: Copied from LayoutTests/fast/css-generated-content/first-letter-in-nested-before-table.html.
* fast/css-generated-content/first-letter-in-nested-before-table-expected.png: Removed.
* fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.
* fast/css-generated-content/first-letter-in-nested-before-table.html:
* fast/css-generated-content/first-letter-in-nested-before.html:

Make reftest and update the behavior. ::first-letter won't affect ::before/after with display:table,
matching Firefox behavior (and spec logic).

* platform/glib/fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.
* platform/gtk/fast/css-generated-content/first-letter-in-nested-before-expected.png: Removed.
* platform/gtk/fast/css-generated-content/first-letter-in-nested-before-table-expected.png: Removed.
* platform/ios/fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.
* platform/mac/fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.
* platform/win/fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.
* platform/wincairo/fast/css-generated-content/first-letter-in-nested-before-table-expected.txt: Removed.

Canonical link: https://commits.webkit.org/250633@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@294291 268f45cc-cd09-0410-ab3c-d52691b4dbfc

rkirsling committed:
ISO8601::parseCalendarTime must throw in cases of time/date ambiguity
https://bugs.webkit.org/show_bug.cgi?id=240498

Reviewed by Yusuke Suzuki.

The CalendarTime production (https://tc39.es/proposal-temporal/#prod-CalendarTime) goes to a lot of trouble
to ensure that ambiguous parses are not allowed; this patch ensures that our implementation does so too.

In short:
- CalendarTime is basically TimeDesignator[opt] TimeSpec TimeZone[opt] Calendar[opt].
- When we don't have a TimeDesignator or a Calendar, we're in danger of ambiguity:
  YYYY-MM, YYYYMM, MM-DD, MMDD look the same as HHMM-UU, HHMMSS, HH-UU, HHMM (where UU is a UTC offset).
- The solution is to attempt to reparse the string as one of these date formats and throw if we succeed.

* stress/temporal-plaintime.js: Adjust expectations.
* test262/expectations.yaml: Mark 10 test cases as passing.

* runtime/ISO8601.cpp:
(JSC::ISO8601::isAmbiguousCalendarTime):
(JSC::ISO8601::parseCalendarTime):
(JSC::ISO8601::daysInMonth):
* runtime/ISO8601.h:

Canonical link: https://commits.webkit.org/250634@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@294306 268f45cc-cd09-0410-ab3c-d52691b4dbfc
